pkgs/development/interpreters/python/cpython/2.7/link-opt.patch

```diff
@@ -0,0 +1,30 @@
+Call the linker with -O1 -Bsymbolic-functions.
+
+Based on debian/patches/link-opt.diff in the Debian package `python3`.
```

Comment:
Can we get a link to salsa and list the changes from the Debian patch?

Reply:
Sure. Here's the patch on salsa (as of latest master):
https://salsa.debian.org/cpython-team/python3/-/blob/02735af38/debian/patches/link-opt.diff

This patch has exactly the same content -- the only change is in the surrounding context.

Similarly, within Debian, the patch's content has been unchanged (the only changes to the patch file have been to update the context and line numbers) since LDCXXSHARED was introduced to the upstream version of this code in Python 3.2 (and 2.7). Here's that patch as it appeared for `python3.2` in Debian 7 Wheezy, released in 2013:
https://sources.debian.org/src/python3.2/3.2.3-7/debian/patches/link-opt.diff/

And here it is from the `python2.4` in Debian 5 Lenny, released in 2009:
https://sources.debian.org/src/python2.4/2.4.6-1+lenny1/debian/patches/link-opt.dpatch/
That one already has the same content except for the absence of LDCXXSHARED.

I'll add into the patch files the Salsa link, and the clarification of "based on" to "has the same content".

Reply:
(Oh, and the Debian versions of the patch apply to `configure.ac`, or on older CPython versions `configure.in`, because Debian makes a point of regenerating `configure` from those.)

Reply:
Regenerating `configure` is not a bad idea either, just add the hook.

Reply:
Oh neat -- searching a bit, I guess that'd be `autoreconfHook`? And we could enable it by just adding it to `nativeBuildInputs`, IIUC. Cool.

Happy to add a commit making that change too. I won't do that tonight, only because testing properly causes rebuilds of GCC etc. that take over an hour, and it's too late in the evening for another round of that.

Reply:
If the patches from Debian apply, it is better to fetch them using `fetchpatch`.

Reply:
> If the patches from Debian apply, it is better to fetch them using `fetchpatch`.

OK, can do.

Any preference on fetching from Salsa vs. sources.debian.org? (The latter being a record of actual releases.)

I think I can find all 4 of the needed versions of the patch (the context keeps changing slightly) on Salsa, and 3 of the 4 on sources -- Debian skipped Python 3.6.

Reply:
Salsa is better since it is a git repository – we can pin it to a commit and it will always be there. Sources might remove older debian releases after they are no longer supported.

Reply:
Mmm, that's a good consideration. On that score, I think actually sources is more dependable -- it isn't the live repository for active releases, rather is meant as a permanent archive for studying long-term history. It has all Debian releases going back to 2.0 "hamm", released in 1998:
https://sources.debian.org/stats/

These patches are probably pretty stable in their URLs on Salsa, too. But they could go away -- the package's maintainer, or a future new maintainer, could move its development to some other repo and remove this one. If it's simply moved within Salsa there'd probably be a redirect, but they could also move it to some other Git hosting service or even something else. Anything the particular maintainer finds convenient for ongoing development would be seen as 100% OK from a Debian perspective.

Reply:
IIRC, sources only carry the latest revision they have for a given Debian release:

https://sources.debian.org/src/python3.2/

Maybe there is some archive site but I do not see revision `-6` in the above link.

Salsa should be stable enough, we can always deal with breakage like we did when they switched from Alioth to Salsa. And Hydra caches the sources so even if they switch, users will still get it from there.

Reply:
> IIRC, sources only carry the latest revision they have for a given Debian release:

Hmm, true. I'd missed that point, thanks. That means nothing will change for any release that's no longer supported, but it can for a release that is (or is still in development).

> https://sources.debian.org/src/python3.2/
> Maybe there is some archive site but I do not see revision `-6` in the above link.

For an even sharper example:
https://sources.debian.org/src/python3.5/

The version available is 3.5.3-1+deb9u1, from Debian 9 "stretch". The "+debXuY" suffix means that was an update after the release was stable, which means the previous version 3.5.3-1 was the one that in the original stable release of "stretch". But that original version doesn't appear on sources. If there's another update to that package in "stretch" in its remaining months before EOL, then presumably the +deb9u1 version will disappear too.

There is a comprehensive archive site which has all versions:
http://snapshot.debian.org/package/python3.2/
But that one only provides whole package files, not individual files from within the package, so it's not suitable for fetchpatch.

I'll use Salsa. Thanks for the discussion!

> And Hydra caches the sources so even if they switch, users will still get it from there.

Oh, good to know. That's quite reassuring for the impact of temporary outages, too.